### ISSUE TITLE
Windows close in the wrong order

### ISSUE THREAD

**Issue by lukepighetti** [helix-editor/helix#4539]
### Summary

If you open a bunch of windows, then start closing them, it always navigates the cursor to the first window, instead of the last selected window. Ideally the focus path travelled during window creation should be followed during window teardown

### Reproduction Steps

 

https://user-images.githubusercontent.com/666539/199011992-a1fa8309-2ed5-4802-a1be-99999412d119.mov

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
please provide a copy of `~/.cache/helix/helix.log` here if possible, you may need to redact some of the lines
```

</details>

### Platform

macOS

### Terminal Emulator

iTerm2

### Helix Version

helix 22.08.1 (6752c7d3)

**Comment by wes-adams** [comment]
i'd like to address this issue

**Comment by wes-adams** [comment]
MR: https://github.com/helix-editor/helix/pull/4766

**Comment by wes-adams** [comment]
**good case:**

https://asciinema.org/a/7RuDbasCErUgVhI3G2BSc14yz

**bad case:**

split view, and then try to close original view

https://asciinema.org/a/l2A7sQfbeGpouIvt5MoNeDzfH

---

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', helix-view/src/tree.rs:286:29
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/option.rs:755:21
   4: helix_view::tree::Tree::get
             at ./helix-view/src/tree.rs:286:9
   5: helix_view::editor::Editor::cursor
             at ./helix-view/src/editor.rs:1316:27
   6: <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::cursor
             at ./helix-term/src/ui/editor.rs:1553:15
   7: helix_term::compositor::Compositor::cursor
             at ./helix-term/src/compositor.rs:173:40
   8: helix_term::application::Application::render
             at ./helix-term/src/application.rs:298:27
   9: helix_term::application::Application::handle_terminal_events
             at ./helix-term/src/application.rs:599:13
  10: helix_term::application::Application::event_loop_until_idle::{{closure}}
             at ./helix-term/src/application.rs:332:21
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  12: helix_term::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:311:57
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  14: helix_term::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:1009:38
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  16: hx::main_impl::{{closure}}
             at ./helix-term/src/main.rs:155:53
  17: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  18: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/park/thread.rs:267:54
  19: tokio::coop::with_budget::{{closure}}
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:102:9
  20: std::thread::local::LocalKey<T>::try_with
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:442:16
  21: std::thread::local::LocalKey<T>::with
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:418:9
  22: tokio::coop::with_budget
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:95:5
  23: tokio::coop::budget
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:72:5
  24: tokio::park::thread::CachedParkThread::block_on
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/park/thread.rs:267:31
  25: tokio::runtime::enter::Enter::block_on
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/enter.rs:152:13
  26: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/scheduler/multi_thread/mod.rs:79:9
  27: tokio::runtime::Runtime::block_on
             at /home/wadams/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/mod.rs:492:44
  28: hx::main_impl
             at ./helix-term/src/main.rs:157:5
  29: hx::main
             at ./helix-term/src/main.rs:37:21
  30: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
  ```

**Comment by wes-adams** [comment]
i think after https://github.com/helix-editor/helix/pull/4766/commits/755c933df322cadb0fbd2e477616ff800521c3f7, the feature is complete

it would be great to get some help testing - is there a typical way to recruit help testing?

**Comment by lukepighetti** [comment]
Can this be modeled as a stack?

open: [0]
split left: [0,1]
split down [0,1,2]
go to 1 [0,2,1]
go to 2 [0, 1, 2]
go to 0 [1, 2, 0]
close [1,2]
close [1]
close quit

**Comment by lukepighetti** [comment]
With the stack approach, if the user selects a window, it is moved to the top of the stack.

**Comment by wes-adams** [comment]
i've re-worked the pull request in a different direction based on feedback from @archseer 
it seems to work. it seems stable. its ready for further review and testing

**Comment by the-mikedavis** [comment]
Closed by #4766
